src/scheduler.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "scheduler.h"
+
+#include <boost/bind.hpp>
+#include <utility>
+
+CScheduler::CScheduler()
+{
+}
+
+
+boost::thread*
+CScheduler::run()
+{
+    boost::thread* t = new boost::thread(boost::bind(&CScheduler::SchedulerLoop, this));
+    return t;
+}
+
+void
+CScheduler::SchedulerLoop()
+{
+    boost::unique_lock<boost::mutex> lock(newTaskMutex);
+
+    // newTaskMutex is locked throughout this loop EXCEPT
+    // when the thread is waiting or when the user's function
+    // is called.
+    while (1)
+    {
+        if (taskQueue.empty())
+        {
+            // Wait until there is something to do.
+            newTaskScheduled.wait(lock);
+        }
+        else
+        {
+            // Wait until either there is a new task, or until
+            // the time of the first item on the queue:
+            while (!taskQueue.empty() &&
+                   newTaskScheduled.wait_until(lock, taskQueue.begin()->first) != boost::cv_status::timeout)
+            {
+                // Keep waiting until timeout
+            }
+            if (taskQueue.empty()) continue;
+
+            Function f = taskQueue.begin()->second;
+            taskQueue.erase(taskQueue.begin());
+
+            // Unlock before calling f, so it can reschedule itself or another task
+            // without deadlocking:
+            lock.unlock();
+            f();
+            lock.lock();
+        }
+    }
+}
+
+void
+CScheduler::schedule(CScheduler::Function f, boost::chrono::system_clock::time_point t)
+{
+    {
+        boost::unique_lock<boost::mutex> lock(newTaskMutex);
+        taskQueue.insert(std::make_pair(t, f));
+    }
+    newTaskScheduled.notify_all();
```

Comment:
Nit: notify_one should be enough (and slightly more efficient, if we'd ever have multiple runner threads).